File: backend/app/models/dunning.py
from __future__ import annotations
import uuid
import enum
from sqlalchemy import (
    Column, String, Integer, Numeric, Boolean, Text, Date, DateTime,
    ForeignKey, Enum, UniqueConstraint, JSON,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from app.db.base import Base, TimestampMixin


# ── Enums ─────────────────────────────────────────────────────────────────────

class DunningActionType(str, enum.Enum):
    EMAIL             = "EMAIL"
    SMS               = "SMS"
    WHATSAPP          = "WHATSAPP"
    TASK              = "TASK"
    CALL_REMINDER     = "CALL_REMINDER"
    CREDIT_HOLD       = "CREDIT_HOLD"
    MANAGER_ESCALATION= "MANAGER_ESCALATION"
    FINAL_NOTICE      = "FINAL_NOTICE"
    LEGAL_REVIEW_FLAG = "LEGAL_REVIEW_FLAG"
    PORTAL_NOTIFICATION = "PORTAL_NOTIFICATION"


class DunningCaseStatus(str, enum.Enum):
    OPEN          = "OPEN"
    ON_HOLD       = "ON_HOLD"
    DISPUTED      = "DISPUTED"
    PROMISE_TO_PAY= "PROMISE_TO_PAY"
    RESOLVED      = "RESOLVED"
    CLOSED        = "CLOSED"
    ESCALATED     = "ESCALATED"


class DunningActionOutcome(str, enum.Enum):
    SENT           = "SENT"
    DELIVERED      = "DELIVERED"
    FAILED         = "FAILED"
    ACKNOWLEDGED   = "ACKNOWLEDGED"
    MANUAL_LOGGED  = "MANUAL_LOGGED"
    SKIPPED        = "SKIPPED"


class PTPStatus(str, enum.Enum):
    OPEN           = "OPEN"
    PARTIALLY_MET  = "PARTIALLY_MET"
    MET            = "MET"
    BROKEN         = "BROKEN"
    CANCELLED      = "CANCELLED"


class DunningExceptionType(str, enum.Enum):
    HOLD_SKIP                  = "HOLD_SKIP"
    CREDIT_HOLD_OVERRIDE       = "CREDIT_HOLD_OVERRIDE"
    DISPUTED_INVOICE_EXCLUSION = "DISPUTED_INVOICE_EXCLUSION"
    COLLECTOR_OVERRIDE         = "COLLECTOR_OVERRIDE"
    CUSTOMER_SPECIAL_TERMS     = "CUSTOMER_SPECIAL_TERMS"


class DunningAIAgentType(str, enum.Enum):
    COLLECTION_PRIORITIZATION = "COLLECTION_PRIORITIZATION"
    PAYMENT_RISK_MONITOR      = "PAYMENT_RISK_MONITOR"
    EFFECTIVENESS_ASSISTANT   = "EFFECTIVENESS_ASSISTANT"


class DunningAIRecStatus(str, enum.Enum):
    PENDING      = "PENDING"
    ACKNOWLEDGED = "ACKNOWLEDGED"
    ACTIONED     = "ACTIONED"
    DISMISSED    = "DISMISSED"


# ── Models ────────────────────────────────────────────────────────────────────

class DunningPolicy(Base, TimestampMixin):
    __tablename__ = "dunning_policies"

    id                          = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    policy_code                 = Column(String(50), nullable=False, unique=True)
    policy_name                 = Column(String(200), nullable=False)
    active_flag                 = Column(Boolean, nullable=False, default=True)
    applies_to_customer_id      = Column(UUID(as_uuid=True), ForeignKey("customers.id"), nullable=True)
    applies_to_channel          = Column(String(60), nullable=True)
    applies_to_region           = Column(String(100), nullable=True)
    applies_to_country          = Column(String(100), nullable=True)
    applies_to_segment          = Column(String(100), nullable=True)
    default_flag                = Column(Boolean, nullable=False, default=False)
    auto_credit_hold_days       = Column(Integer, nullable=True)
    auto_credit_hold_amount     = Column(Numeric(16, 2), nullable=True)
    notes                       = Column(Text, nullable=True)

    levels = relationship("DunningLevel", back_populates="policy", cascade="all, delete-orphan", order_by="DunningLevel.level_no")


class DunningLevel(Base, TimestampMixin):
    __tablename__ = "dunning_levels"
    __table_args__ = (UniqueConstraint("policy_id", "level_no", name="uq_dunning_level"),)

    id                    = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    policy_id             = Column(UUID(as_uuid=True), ForeignKey("dunning_policies.id"), nullable=False)
    level_no              = Column(Integer, nullable=False)
    level_name            = Column(String(100), nullable=False)
    days_overdue_from     = Column(Integer, nullable=False)
    days_overdue_to       = Column(Integer, nullable=True)
    action_type           = Column(Enum(DunningActionType), nullable=False, default=DunningActionType.EMAIL)
    template_code         = Column(String(100), nullable=True)
    auto_send_flag        = Column(Boolean, nullable=False, default=True)
    requires_approval_flag= Column(Boolean, nullable=False, default=False)
    apply_credit_hold_flag= Column(Boolean, nullable=False, default=False)
    escalate_to_role      = Column(String(100), nullable=True)
    active_flag           = Column(Boolean, nullable=False, default=True)
    notes                 = Column(Text, nullable=True)

    policy = relationship("DunningPolicy", back_populates="levels")


class DunningTemplate(Base, TimestampMixin):
    __tablename__ = "dunning_templates"
    __table_args__ = (UniqueConstraint("template_code", "language", name="uq_dunning_template"),)

    id            = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    template_code = Column(String(100), nullable=False)
    language      = Column(String(10), nullable=False, default="en")
    subject       = Column(String(300), nullable=False)
    body_text     = Column(Text, nullable=False)
    channel       = Column(Enum(DunningActionType), nullable=False, default=DunningActionType.EMAIL)
    active_flag   = Column(Boolean, nullable=False, default=True)
    notes         = Column(Text, nullable=True)


class DunningCase(Base, TimestampMixin):
    __tablename__ = "dunning_cases"

    id                       = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    case_no                  = Column(String(50), nullable=False, unique=True)
    customer_id              = Column(UUID(as_uuid=True), ForeignKey("customers.id"), nullable=False)
    policy_id                = Column(UUID(as_uuid=True), ForeignKey("dunning_policies.id"), nullable=True)
    current_level_no         = Column(Integer, nullable=False, default=0)
    case_status              = Column(Enum(DunningCaseStatus), nullable=False, default=DunningCaseStatus.OPEN)
    total_overdue_amount     = Column(Numeric(16, 2), nullable=False, default=0)
    oldest_due_date          = Column(Date, nullable=True)
    latest_action_date       = Column(DateTime(timezone=True), nullable=True)
    next_action_date         = Column(Date, nullable=True)
    assigned_collector_id    = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    credit_hold_flag         = Column(Boolean, nullable=False, default=False)
    credit_hold_reason       = Column(Text, nullable=True)
    credit_hold_date         = Column(Date, nullable=True)
    priority_score           = Column(Integer, nullable=False, default=0)
    collector_notes          = Column(Text, nullable=True)
    notes                    = Column(Text, nullable=True)

    invoice_links  = relationship("DunningCaseInvoice", back_populates="case", cascade="all, delete-orphan")
    action_logs    = relationship("DunningActionLog", back_populates="case", cascade="all, delete-orphan")
    ptps           = relationship("DunningPTP", back_populates="case", cascade="all, delete-orphan")
    exceptions     = relationship("DunningException", back_populates="case", cascade="all, delete-orphan")


class DunningCaseInvoice(Base, TimestampMixin):
    __tablename__ = "dunning_case_invoices"
    __table_args__ = (UniqueConstraint("case_id", "invoice_id", name="uq_dunning_case_inv"),)

    id                      = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    case_id                 = Column(UUID(as_uuid=True), ForeignKey("dunning_cases.id"), nullable=False)
    invoice_id              = Column(UUID(as_uuid=True), ForeignKey("invoices.id"), nullable=False)
    invoice_number          = Column(String(80), nullable=False)
    invoice_due_date        = Column(Date, nullable=False)
    invoice_balance_amount  = Column(Numeric(16, 2), nullable=False)
    days_overdue            = Column(Integer, nullable=False, default=0)
    dispute_flag            = Column(Boolean, nullable=False, default=False)
    dispute_reason          = Column(Text, nullable=True)
    promise_to_pay_date     = Column(Date, nullable=True)
    excluded_from_dunning   = Column(Boolean, nullable=False, default=False)
    notes                   = Column(Text, nullable=True)

    case = relationship("DunningCase", back_populates="invoice_links")


class DunningActionLog(Base):
    __tablename__ = "dunning_action_logs"

    id                  = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    case_id             = Column(UUID(as_uuid=True), ForeignKey("dunning_cases.id"), nullable=False)
    invoice_id          = Column(UUID(as_uuid=True), nullable=True)
    level_no            = Column(Integer, nullable=False)
    action_type         = Column(Enum(DunningActionType), nullable=False)
    channel_type        = Column(String(40), nullable=True)
    action_datetime     = Column(DateTime(timezone=True), nullable=False)
    sent_by_system_flag = Column(Boolean, nullable=False, default=True)
    user_id             = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    template_code       = Column(String(100), nullable=True)
    outcome_status      = Column(Enum(DunningActionOutcome), nullable=False, default=DunningActionOutcome.SENT)
    external_message_ref= Column(String(200), nullable=True)
    rendered_subject    = Column(String(400), nullable=True)
    rendered_body       = Column(Text, nullable=True)
    notes               = Column(Text, nullable=True)

    case = relationship("DunningCase", back_populates="action_logs")


class DunningPTP(Base, TimestampMixin):
    __tablename__ = "dunning_ptps"

    id                   = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    case_id              = Column(UUID(as_uuid=True), ForeignKey("dunning_cases.id"), nullable=False)
    customer_id          = Column(UUID(as_uuid=True), ForeignKey("customers.id"), nullable=False)
    promised_amount      = Column(Numeric(16, 2), nullable=False)
    promised_date        = Column(Date, nullable=False)
    recorded_by          = Column(String(200), nullable=False)
    recorded_at          = Column(DateTime(timezone=True), nullable=False)
    status               = Column(Enum(PTPStatus), nullable=False, default=PTPStatus.OPEN)
    notes                = Column(Text, nullable=True)

    case = relationship("DunningCase", back_populates="ptps")


class DunningException(Base, TimestampMixin):
    __tablename__ = "dunning_exceptions"

    id               = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    case_id          = Column(UUID(as_uuid=True), ForeignKey("dunning_cases.id"), nullable=False)
    exception_type   = Column(Enum(DunningExceptionType), nullable=False)
    approved_by      = Column(String(200), nullable=False)
    start_date       = Column(Date, nullable=False)
    end_date         = Column(Date, nullable=True)
    reason           = Column(Text, nullable=False)
    notes            = Column(Text, nullable=True)

    case = relationship("DunningCase", back_populates="exceptions")


class DunningAIRecommendation(Base, TimestampMixin):
    __tablename__ = "dunning_ai_recommendations"

    id             = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    case_id        = Column(UUID(as_uuid=True), ForeignKey("dunning_cases.id"), nullable=True)
    customer_id    = Column(UUID(as_uuid=True), ForeignKey("customers.id"), nullable=True)
    agent_type     = Column(Enum(DunningAIAgentType), nullable=False)
    status         = Column(Enum(DunningAIRecStatus), nullable=False, default=DunningAIRecStatus.PENDING)
    title          = Column(String(300), nullable=False)
    body           = Column(Text, nullable=False)
    severity       = Column(String(20), nullable=False, default="INFO")
    reference_type = Column(String(80), nullable=True)
    reference_id   = Column(String(80), nullable=True)
    actioned_by    = Column(String(200), nullable=True)
    actioned_at    = Column(DateTime(timezone=True), nullable=True)

File: frontend/src/lib/dunning.ts
const BASE = `${process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000"}/api/v1/dunning`;

async function r<T>(res: Response): Promise<T> {
  if (!res.ok) throw new Error(await res.text());
  return res.json();
}

// ── Types ─────────────────────────────────────────────────────────────────────

export type DunningActionType = "EMAIL" | "SMS" | "WHATSAPP" | "TASK" | "CALL_REMINDER" | "CREDIT_HOLD" | "MANAGER_ESCALATION" | "FINAL_NOTICE" | "LEGAL_REVIEW_FLAG" | "PORTAL_NOTIFICATION";
export type DunningCaseStatus = "OPEN" | "ON_HOLD" | "DISPUTED" | "PROMISE_TO_PAY" | "RESOLVED" | "CLOSED" | "ESCALATED";
export type DunningActionOutcome = "SENT" | "DELIVERED" | "FAILED" | "ACKNOWLEDGED" | "MANUAL_LOGGED" | "SKIPPED";
export type PTPStatus = "OPEN" | "PARTIALLY_MET" | "MET" | "BROKEN" | "CANCELLED";
export type DunningExceptionType = "HOLD_SKIP" | "CREDIT_HOLD_OVERRIDE" | "DISPUTED_INVOICE_EXCLUSION" | "COLLECTOR_OVERRIDE" | "CUSTOMER_SPECIAL_TERMS";
export type DunningAIAgentType = "COLLECTION_PRIORITIZATION" | "PAYMENT_RISK_MONITOR" | "EFFECTIVENESS_ASSISTANT";
export type DunningAIRecStatus = "PENDING" | "ACKNOWLEDGED" | "ACTIONED" | "DISMISSED";

export interface DunningPolicy {
  id: string;
  policy_code: string;
  policy_name: string;
  active_flag: boolean;
  default_flag: boolean;
  applies_to_channel?: string;
  applies_to_region?: string;
  applies_to_segment?: string;
  auto_credit_hold_days?: number;
  auto_credit_hold_amount?: number;
  notes?: string;
  created_at?: string;
}

export interface DunningLevel {
  id: string;
  policy_id: string;
  level_no: number;
  level_name: string;
  days_overdue_from: number;
  days_overdue_to?: number;
  action_type: DunningActionType;
  template_code?: string;
  auto_send_flag: boolean;
  requires_approval_flag: boolean;
  apply_credit_hold_flag: boolean;
  escalate_to_role?: string;
  active_flag: boolean;
}

export interface DunningTemplate {
  id: string;
  template_code: string;
  language: string;
  subject: string;
  body_text: string;
  channel: DunningActionType;
  active_flag: boolean;
  notes?: string;
}

export interface DunningCase {
  id: string;
  case_no: string;
  customer_id: string;
  policy_id?: string;
  current_level_no: number;
  case_status: DunningCaseStatus;
  total_overdue_amount: number;
  oldest_due_date?: string;
  latest_action_date?: string;
  next_action_date?: string;
  assigned_collector_id?: string;
  credit_hold_flag: boolean;
  credit_hold_reason?: string;
  credit_hold_date?: string;
  priority_score: number;
  collector_notes?: string;
  notes?: string;
  created_at?: string;
}

export interface DunningCaseInvoice {
  id: string;
  case_id: string;
  invoice_id: string;
  invoice_number: string;
  invoice_due_date: string;
  invoice_balance_amount: number;
  days_overdue: number;
  dispute_flag: boolean;
  dispute_reason?: string;
  promise_to_pay_date?: string;
  excluded_from_dunning: boolean;
}

export interface DunningActionLog {
  id: string;
  case_id: string;
  level_no: number;
  action_type: DunningActionType;
  action_datetime: string;
  sent_by_system_flag: boolean;
  outcome_status: DunningActionOutcome;
  rendered_subject?: string;
  rendered_body?: string;
  notes?: string;
}

export interface DunningPTP {
  id: string;
  case_id: string;
  customer_id: string;
  promised_amount: number;
  promised_date: string;
  recorded_by: string;
  recorded_at: string;
  status: PTPStatus;
  notes?: string;
}

export interface DunningException {
  id: string;
  case_id: string;
  exception_type: DunningExceptionType;
  approved_by: string;
  start_date: string;
  end_date?: string;
  reason: string;
}

export interface DunningAIRec {
  id: string;
  case_id?: string;
  customer_id?: string;
  agent_type: DunningAIAgentType;
  status: DunningAIRecStatus;
  title: string;
  body: string;
  severity: string;
  reference_type?: string;
  reference_id?: string;
  actioned_by?: string;
  created_at?: string;
}

export interface AgingRow {
  customer_id: string;
  customer_name: string;
  customer_code: string;
  currency: string;
  current: number;
  overdue_1_7: number;
  overdue_8_14: number;
  overdue_15_30: number;
  overdue_31_60: number;
  overdue_61_90: number;
  overdue_90_plus: number;
  total_overdue: number;
  credit_hold: boolean;
  case_id?: string;
  priority_score: number;
}

export interface DunningDashboard {
  open_cases: number;
  total_overdue_amount: number;
  credit_hold_count: number;
  disputed_cases: number;
  open_ptps: number;
  broken_ptps: number;
}

// ── Color Maps ────────────────────────────────────────────────────────────────

export const CASE_STATUS_COLORS: Record<DunningCaseStatus, string> = {
  OPEN: "bg-blue-100 text-blue-800",
  ON_HOLD: "bg-yellow-100 text-yellow-800",
  DISPUTED: "bg-orange-100 text-orange-800",
  PROMISE_TO_PAY: "bg-purple-100 text-purple-800",
  RESOLVED: "bg-green-100 text-green-800",
  CLOSED: "bg-gray-100 text-gray-600",
  ESCALATED: "bg-red-100 text-red-800",
};

export const PTP_STATUS_COLORS: Record<PTPStatus, string> = {
  OPEN: "bg-blue-100 text-blue-800",
  PARTIALLY_MET: "bg-yellow-100 text-yellow-800",
  MET: "bg-green-100 text-green-800",
  BROKEN: "bg-red-100 text-red-800",
  CANCELLED: "bg-gray-100 text-gray-600",
};

export const OUTCOME_COLORS: Record<DunningActionOutcome, string> = {
  SENT: "bg-blue-100 text-blue-800",
  DELIVERED: "bg-green-100 text-green-800",
  FAILED: "bg-red-100 text-red-800",
  ACKNOWLEDGED: "bg-teal-100 text-teal-800",
  MANUAL_LOGGED: "bg-gray-100 text-gray-700",
  SKIPPED: "bg-gray-100 text-gray-500",
};

export const SEVERITY_COLORS: Record<string, string> = {
  INFO: "bg-blue-100 text-blue-800",
  WARNING: "bg-yellow-100 text-yellow-800",
  CRITICAL: "bg-red-100 text-red-800",
};

// ── API Client ────────────────────────────────────────────────────────────────

const j = { "Content-Type": "application/json" };
const post = (url: string, body: any) => fetch(url, { method: "POST", headers: j, body: JSON.stringify(body) });
const patch = (url: string, body: any) => fetch(url, { method: "PATCH", headers: j, body: JSON.stringify(body) });
const put = (url: string, body: any) => fetch(url, { method: "PUT", headers: j, body: JSON.stringify(body) });

export const dunningApi = {
  // Policies
  listPolicies: (activeOnly = false) => fetch(`${BASE}/policies?active_only=${activeOnly}`).then(r<DunningPolicy[]>),
  createPolicy: (body: Partial<DunningPolicy>) => post(`${BASE}/policies`, body).then(r<DunningPolicy>),
  updatePolicy: (id: string, body: Partial<DunningPolicy>) => put(`${BASE}/policies/${id}`, body).then(r<DunningPolicy>),
  listLevels: (policyId: string) => fetch(`${BASE}/policies/${policyId}/levels`).then(r<DunningLevel[]>),
  addLevel: (policyId: string, body: Partial<DunningLevel>) => post(`${BASE}/policies/${policyId}/levels`, body).then(r<DunningLevel>),

  // Templates
  listTemplates: () => fetch(`${BASE}/templates`).then(r<DunningTemplate[]>),
  createTemplate: (body: Partial<DunningTemplate>) => post(`${BASE}/templates`, body).then(r<DunningTemplate>),

  // Detection
  runDetection: () => post(`${BASE}/run-overdue-detection`, {}).then(r<any>),
  checkBrokenPTPs: () => post(`${BASE}/check-broken-ptps`, {}).then(r<any>),

  // Cases
  listCases: (params?: { status?: string; credit_hold?: boolean; limit?: number }) => {
    const p = new URLSearchParams();
    if (params?.status) p.set("status", params.status);
    if (params?.credit_hold !== undefined) p.set("credit_hold", String(params.credit_hold));
    if (params?.limit) p.set("limit", String(params.limit));
    return fetch(`${BASE}/cases?${p}`).then(r<DunningCase[]>);
  },
  getCase: (id: string) => fetch(`${BASE}/cases/${id}`).then(r<DunningCase>),
  updateNotes: (id: string, collector_notes: string) => patch(`${BASE}/cases/${id}/notes`, { collector_notes }).then(r<DunningCase>),
  assignCase: (id: string, collector_id: string) => post(`${BASE}/cases/${id}/assign`, { assigned_collector_id: collector_id }).then(r<DunningCase>),
  closeCase: (id: string) => post(`${BASE}/cases/${id}/close`, {}).then(r<DunningCase>),

  // Case Invoices
  getCaseInvoices: (caseId: string) => fetch(`${BASE}/cases/${caseId}/invoices`).then(r<DunningCaseInvoice[]>),
  setDispute: (caseId: string, body: { invoice_id: string; dispute_flag: boolean; dispute_reason?: string }) =>
    post(`${BASE}/cases/${caseId}/dispute`, body).then(r<DunningCaseInvoice>),

  // Actions
  sendReminder: (caseId: string, channel: DunningActionType, note?: string) =>
    post(`${BASE}/cases/${caseId}/send-reminder`, { channel, manual_note: note }).then(r<DunningActionLog>),
  addNote: (caseId: string, notes: string) =>
    post(`${BASE}/cases/${caseId}/add-note`, { notes }).then(r<DunningActionLog>),
  listActions: (caseId: string) => fetch(`${BASE}/cases/${caseId}/actions`).then(r<DunningActionLog[]>),

  // PTPs
  listPTPs: (caseId: string) => fetch(`${BASE}/cases/${caseId}/ptps`).then(r<DunningPTP[]>),
  createPTP: (caseId: string, body: { promised_amount: number; promised_date: string; recorded_by: string; notes?: string }) =>
    post(`${BASE}/cases/${caseId}/set-promise-to-pay`, body).then(r<DunningPTP>),
  updatePTP: (ptpId: string, status: PTPStatus, notes?: string) =>
    patch(`${BASE}/ptps/${ptpId}`, { status, notes }).then(r<DunningPTP>),

  // Exceptions
  listExceptions: (caseId: string) => fetch(`${BASE}/cases/${caseId}/exceptions`).then(r<DunningException[]>),
  createException: (caseId: string, body: Partial<DunningException>) =>
    post(`${BASE}/cases/${caseId}/exceptions`, body).then(r<DunningException>),

  // Credit Hold
  applyCreditHold: (caseId: string, reason: string, applied_by: string) =>
    post(`${BASE}/cases/${caseId}/apply-credit-hold`, { reason, applied_by }).then(r<DunningCase>),
  releaseCreditHold: (caseId: string, released_by: string, notes?: string) =>
    post(`${BASE}/cases/${caseId}/release-credit-hold`, { released_by, notes }).then(r<DunningCase>),

  // AI
  runAI: () => post(`${BASE}/ai/run`, {}).then(r<{ generated: number }>),
  listAIRecs: (caseId?: string) =>
    fetch(`${BASE}/ai/recommendations${caseId ? `?case_id=${caseId}` : ""}`).then(r<DunningAIRec[]>),
  ackAIRec: (recId: string, status: DunningAIRecStatus, actioned_by: string) =>
    post(`${BASE}/ai/recommendations/${recId}/ack`, { status, actioned_by }).then(r<DunningAIRec>),

  // Reports
  getDashboard: () => fetch(`${BASE}/reports/dashboard`).then(r<DunningDashboard>),
  getAging: () => fetch(`${BASE}/reports/aging`).then(r<AgingRow[]>),
  getEffectiveness: () => fetch(`${BASE}/reports/effectiveness`).then(r<any[]>),
  getWorkqueue: (collectorId?: string) =>
    fetch(`${BASE}/reports/workqueue${collectorId ? `?collector_id=${collectorId}` : ""}`).then(r<any[]>),
  getPTPFulfillment: () => fetch(`${BASE}/reports/ptp-fulfillment`).then(r<any>),
};
